src/App/Security/Authorization/Voter/ArticleVoter.php

```diff
@@ -38,34 +46,72 @@ protected function supports($attribute, $subject)
 
     protected function voteOnAttribute($attribute, $subject, TokenInterface $token)
     {
+        $article = $subject;
+
+        switch ($attribute) {
+            case self::PUBLISHED:
+                return $this->isPublished($article);
+            case self::LOCKED:
+                return $this->isLocked($article);
+            case self::REMOVED:
+                return $this->isRemoved($article);
+        }
+
         $user = $token->getUser();
 
         if (!$user instanceof User) {
             return false;
         }
 
-        $article = $subject;
-
         switch ($attribute) {
-            case self::EDIT:
-                return $this->canEdit($article, $user);
-            case self::DELETE:
-                return $this->canDelete($article, $user, $token);
+            case self::CAN_BE_EDITED:
+                return $this->canBeEdited($article, $user);
+            case self::CAN_BE_REMOVED:
+                return $this->canBeRemoved($article, $user);
+            case self::SHOW_ADMIN_PANEL:
+                return $this->showAdminPanel($article, $user);
         }
 
         throw new \LogicException('This code should not be reached!');
     }
 
-    private function canEdit($article, $user)
+    private function canBeEdited($article, $user)
     {
-        // We should have the case of a wiki here
+        if (!$this->isPublished($article)) {
+            return false;
+        }
+
         return $user === $article->getProposal()->getAuthor();
     }
 
-    private function canDelete($article, $user, $token)
+    private function canBeRemoved($article, $user)
     {
-        if ($this->decisionManager->decide($token, ['ROLE_ADMIN'])) {
-            return true;
+        if (!$article->isPublished()) {
+            return false;
+        }
+
+        return $user === $article->getProposal()->getAuthor();
+    }
+
+    private function isPublished($article)
+    {
+        return $article->getStatus() == $article::PUBLISHED;
```

Comment:
Penser à utiliser des `===` plutôt que des `==` partout

Reply:
done.

J'ai pas compris le dernier commentaire. "strict type modes" sur les fichiers ?

Merci beaucoup pour les corretions !! :))

Reply:
https://stackoverflow.com/questions/48723637/what-do-strict-types-do-in-php

Reply:
Pareil que pour les type hints. Je suis d'accord que c'est quelque chose à faire mais que je préfèrerais repousser pour l'instant. Ça demandera sûrement beaucoup de travail et je préférerais dédier une PR à ce sujet.

Reply:
I opened an issue for that.